text/0000-mir.md

```diff
@@ -0,0 +1,821 @@
+If this code were written using `p = &mut vec[0], q = &mut vec[1]`,
+the borrow checker would complain. This is because it does not attempt
+to reason about indices being disjoint, even if they are constant
+(this is a limitation we may wish to consider lifting at some point in
+the future, however).
+
+To accommodate these, we plan to desugar such matches into lvalues
+using the special "constant index" form. The borrow checker would be
+able to reason that two constant indices are disjoint but it could
+consider "variable indices" to be (potentially) overlapping with all
+constant indices. This is a fairly straightforward thing to do (and in
+fact the borrow checker already includes similar logic, since the
+`ExprUseVisitor` encounters a similar dilemna trying to resolve
+borrows).
+
+### Drops
+
+The `Drop(DROP_KIND, LVALUE)` instruction is intended to represent
+"automatic" compiler-inserted drops. The semantics of a `Drop` is that
+it drops "if needed". This means that the compiler can insert it
+everywhere that a `Drop` would make sense (due to scoping), and assume
+that instrumentation will be done as needed to prevent double
+drops. Currently, this signaling is done by zeroing out memory at
+runtime, but we are in the process of introducing stack flags for this
+purpose: the MIR offers the opportunity to reify those flags if we
+wanted, and rewrite drops to be more narrow (versus leaving that work
+for LLVM).
+
+To illustrate how drop works, let's work through a simple
+example. Imagine that we have a snippet of code like:
+
+```rust
+{
+  let x = Box::new(22);
+  send(x);
+}
+```
+
+The compiler would generate a drop for `x` at the end of the block,
+but the value `x` would also be moved as part of the call to `send`.
+A later analysis could easily strip out this `Drop` since it is evident
+that the value is always used on all paths that lead to `Drop`.
+
+### Shallow drops and Box
+
+The MIR includes the distinction between "shallow" and "deep"
+drop. Deep drop is the normal thing, but shallow drop is used when
+partially initializing boxes. This is tied to the `box` keyword.
+For example, an assignment like the following:
+
+    let x = box Foo::new();
+    
+would be translated to something like the following:
+
+    let tmp: Box<Foo>;
+    
+    B0: {
+      tmp = BOX;
+      f = Foo::new; // constant reference
+      call(*tmp, f, B1, B2);
+    } 
+    
+    B1: { // successful return of the call
+      x = use(tmp); // move of tmp
+      ...
+    }
+    
+    B2: { // calling Foo::new() panic'd
+      drop(Shallow, tmp);
+      diverge;
+    }
+
+The interesting part here is the block B2, which indicates the case
+that `Foo::new()` invoked unwinding. In that case, we have to free the
+box that we allocated, but we only want to free the box itself, not
+its contents (it is not yet initialized).
+
+Note that having this kind of builtin box code is a legacy thing. The
+more generalized protocol that [RFC 809][809] specifies works in
```

Comment:
I think you meant to link to [RFC 809](https://github.com/rust-lang/rfcs/pull/809) here.